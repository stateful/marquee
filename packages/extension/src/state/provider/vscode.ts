import type { ExtensionContext } from 'vscode'
import type { StateProvider } from '../types'

export class VSCodeState implements StateProvider {
  constructor (public context: ExtensionContext) {}

  getState (key: string) {
    return this.context.globalState.get(key) as never
  }

  async setState (key: string, value: any) {
    await this.context.globalState.update(key, value)
  }
}
