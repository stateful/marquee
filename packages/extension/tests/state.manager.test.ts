import vscode from 'vscode'
import { TextEncoder, TextDecoder } from 'util'

// @ts-expect-error mock
import manager from 'disposableManager'

import StateManager from '../src/stateManager'
import oldConfigFile from './__fixtures__/oldMarqueeConfig.json'

jest.mock('../src/utils', () => ({ activateGUI: () => ({ marquee: { disposable: require('disposableManager') }}) }))
jest.mock('@vscode-marquee/utils/extension', () => ({
  activate: () => ({ marquee: { disposable: require('disposableManager') }}),
  pkg: { version: '1.2.3' },
  DEPRECATED_GLOBAL_STORE_KEY: 'foobar'
}))
jest.mock('@vscode-marquee/widget-welcome/extension', () => (
  { activate: () => ({ marquee: { disposable: require('disposableManager') }}) })
)
jest.mock('@vscode-marquee/widget-projects/extension', () => (
  { activate: () => ({ marquee: { disposable: require('disposableManager') }}) })
)
jest.mock('@vscode-marquee/widget-github/extension', () => (
  { activate: () => ({ marquee: { disposable: require('disposableManager') }}) })
)
jest.mock('@vscode-marquee/widget-weather/extension', () => (
  { activate: () => ({ marquee: { disposable: require('disposableManager') }}) })
)
jest.mock('@vscode-marquee/widget-todo/extension', () => (
  { activate: () => ({ marquee: { disposable: require('disposableManager') }}) })
)
jest.mock('@vscode-marquee/widget-notes/extension', () => (
  { activate: () => ({ marquee: { disposable: require('disposableManager') }}) })
)
jest.mock('@vscode-marquee/widget-snippets/extension', () => (
  { activate: () => ({ marquee: { disposable: require('disposableManager') }}) })
)
jest.mock('@vscode-marquee/widget-markdown/extension', () => (
  { activate: () => ({ marquee: { disposable: require('disposableManager') }}) })
)
jest.mock('@vscode-marquee/widget-npm-stats/extension', () => (
  { activate: () => ({ marquee: { disposable: require('disposableManager') }}) })
)

const encoder = new TextEncoder()
const decoder = new TextDecoder();
(vscode.workspace.fs.readFile as jest.Mock).mockResolvedValue(encoder.encode(JSON.stringify({ foo: 'bar' })));
(vscode.workspace.fs.writeFile as jest.Mock).mockResolvedValue({})

const context = {
  globalState: { update: jest.fn() }
} as any

beforeEach(() => {
  (vscode.window.showInformationMessage as jest.Mock).mockClear()
  context.globalState.update.mockClear()
  manager.setImportInProgress.mockClear()
})

test('StateManager initiation', () => {
  const stateManager = new StateManager(context, 'channel' as any)
  expect(stateManager['_subscriptions']).toMatchSnapshot()
  expect(context.globalState.update).toBeCalledWith('foobar', undefined)
})

test('_import', async () => {
  const stateManager = new StateManager(context, 'channel' as any)
  await stateManager['_import']()
  expect(vscode.window.showErrorMessage).toBeCalledWith(
    'Error importing file: Invalid Marquee Configuration'
  )
  expect(manager.setImportInProgress).toBeCalledTimes(22)
})

test('_import transforms old config types', async () => {
  const stateManager = new StateManager({ ...context, extensionPath: '/foo/bar' } as any, 'channel' as any);
  (vscode.window.showOpenDialog as jest.Mock).mockResolvedValue([{ fsPath: '/foo/bar' }]);
  (vscode.workspace.fs.readFile as jest.Mock).mockResolvedValue(encoder.encode(JSON.stringify(oldConfigFile)))
  await stateManager['_import']()
  expect(manager.updateConfiguration.mock.calls).toMatchSnapshot()
  expect(manager.updateState.mock.calls).toMatchSnapshot()
  expect(manager.emit).toBeCalledWith('gui.close')
  expect(manager.emit).toBeCalledWith('gui.open', true)
  expect(manager.setImportInProgress).toBeCalledTimes(22)
})

test('_export', async () => {
  const stateManager = new StateManager({ ...context, extensionPath: '/foo/bar' } as any, 'channel' as any);
  (vscode.window.showOpenDialog as jest.Mock).mockResolvedValue([{ fsPath: '/foo/bar' }])
  await stateManager['_export']()
  expect(decoder.decode((vscode.workspace.fs.writeFile as jest.Mock).mock.calls[0][1])).toMatchSnapshot()
  expect(vscode.window.showInformationMessage).toBeCalledWith('Successfully exported Marquee state to /bar/foo')
})

test('_export with error', async () => {
  const stateManager = new StateManager({ ...context, extensionPath: '/foo/bar' } as any, 'channel' as any);
  (vscode.window.showOpenDialog as jest.Mock).mockResolvedValue([{ fsPath: '/foo/bar' }]);
  (vscode.workspace.fs.writeFile as jest.Mock).mockRejectedValue(new Error('upps'))
  await stateManager['_export']()
  expect(vscode.window.showErrorMessage).toBeCalledWith('Error writing export file: upps')
})

test('access to individual disposables', () => {
  const stateManager = new StateManager({ ...context, extensionPath: '/foo/bar' } as any, 'channel' as any)
  expect(stateManager.projectWidget).toEqual(manager)
  expect(stateManager.notesWidget).toEqual(manager)
  expect(stateManager.todoWidget).toEqual(manager)
  expect(stateManager.snippetsWidget).toEqual(manager)
  expect(stateManager.global).toEqual(manager)
  expect(stateManager.gui).toEqual(manager)
})

test('clearAll', async () => {
  const stateManager = new StateManager({ ...context, extensionPath: '/foo/bar' } as any, 'channel' as any)
  await stateManager.clearAll()
  expect(manager.clear).toBeCalledTimes(11)
})

test('onWidget', () => {
  const stateManager = new StateManager({ ...context, extensionPath: '/foo/bar' } as any, 'channel' as any)
  stateManager.onWidget('foobar', () => {})
  expect(manager.on).toBeCalledTimes(11)
  expect(manager.on).toBeCalledWith('foobar', expect.any(Function))
})

test('dispose', async () => {
  const stateManager = new StateManager({ ...context, extensionPath: '/foo/bar' } as any, 'channel' as any)
  stateManager['_subscriptions'] = [{ dispose: jest.fn() }]
  await stateManager.dispose()
  expect(stateManager['_subscriptions'][0].dispose).toBeCalledTimes(1)
})
