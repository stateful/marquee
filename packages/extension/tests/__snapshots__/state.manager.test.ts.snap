// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`_export 1`] = `
"{
 \\"type\\": \\"MarqueeSettings\\",
 \\"version\\": \\"1.2.3\\",
 \\"state\\": {
  \\"@vscode-marquee/utils\\": \\"some state\\",
  \\"@vscode-marquee/gui\\": \\"some state\\",
  \\"@vscode-marquee/welcome-widget\\": \\"some state\\",
  \\"@vscode-marquee/projects-widget\\": \\"some state\\",
  \\"@vscode-marquee/github-widget\\": \\"some state\\",
  \\"@vscode-marquee/weather-widget\\": \\"some state\\",
  \\"@vscode-marquee/todo-widget\\": \\"some state\\",
  \\"@vscode-marquee/notes-widget\\": \\"some state\\",
  \\"@vscode-marquee/snippets-widget\\": \\"some state\\"
 },
 \\"configuration\\": {
  \\"@vscode-marquee/utils\\": \\"some configuration\\",
  \\"@vscode-marquee/gui\\": \\"some configuration\\",
  \\"@vscode-marquee/welcome-widget\\": \\"some configuration\\",
  \\"@vscode-marquee/projects-widget\\": \\"some configuration\\",
  \\"@vscode-marquee/github-widget\\": \\"some configuration\\",
  \\"@vscode-marquee/weather-widget\\": \\"some configuration\\",
  \\"@vscode-marquee/todo-widget\\": \\"some configuration\\",
  \\"@vscode-marquee/notes-widget\\": \\"some configuration\\",
  \\"@vscode-marquee/snippets-widget\\": \\"some configuration\\"
 }
}"
`;

exports[`_import transforms old config types 1`] = `
Array [
  Array [
    "name",
    undefined,
  ],
  Array [
    "background",
    "9",
  ],
  Array [
    "language",
    undefined,
  ],
  Array [
    "since",
    "Weekly",
  ],
  Array [
    "spoken",
    undefined,
  ],
  Array [
    "autoDetect",
    undefined,
  ],
]
`;

exports[`_import transforms old config types 2`] = `
Array [
  Array [
    "read",
    Array [
      "-MumT6xMRoa9-5HldrZM",
      "-MsgFT_sLhRzijQekM8s",
      "-MRQu2va0eCaPZnPAtNz",
      "-MO3AlnxSsSx16kAWodv",
      "-MKpao5VyK0UrufFJT6M",
    ],
  ],
  Array [
    "workspaces",
    Array [],
  ],
  Array [
    "todos",
    Array [
      Object {
        "archived": false,
        "body": "foobar",
        "checked": false,
        "exists": false,
        "id": "z825fdsi",
        "tags": Array [
          "foo",
        ],
        "workspaceId": "00824ce3c6d29b87c30ed9649a23cf6ce9554c80",
      },
    ],
  ],
  Array [
    "notes",
    Array [
      Object {
        "archived": false,
        "body": "<p>asdsaasd</p>",
        "createdAt": 1644402822591,
        "id": "eb9s48rn",
        "origin": null,
        "text": "",
        "title": "sadsda",
        "workspaceId": "00824ce3c6d29b87c30ed9649a23cf6ce9554c80",
      },
    ],
  ],
  Array [
    "snippets",
    Array [
      Object {
        "archived": false,
        "body": "dassaddsa",
        "createdAt": 1644402818003,
        "exists": false,
        "id": "y7utrxy7",
        "language": Object {
          "name": "markdown",
          "value": "markdown",
        },
        "origin": null,
        "title": "dassadsad",
        "workspaceId": "00824ce3c6d29b87c30ed9649a23cf6ce9554c80",
      },
    ],
  ],
]
`;

exports[`StateManager initiation 1`] = `
Array [
  Object {
    "clear": [MockFunction],
    "configuration": "some configuration",
    "dispose": [MockFunction],
    "emit": [MockFunction],
    "on": [MockFunction],
    "setImportInProgress": [MockFunction],
    "state": "some state",
    "updateConfiguration": [MockFunction],
    "updateState": [MockFunction],
  },
  Object {
    "clear": [MockFunction],
    "configuration": "some configuration",
    "dispose": [MockFunction],
    "emit": [MockFunction],
    "on": [MockFunction],
    "setImportInProgress": [MockFunction],
    "state": "some state",
    "updateConfiguration": [MockFunction],
    "updateState": [MockFunction],
  },
  Object {
    "clear": [MockFunction],
    "configuration": "some configuration",
    "dispose": [MockFunction],
    "emit": [MockFunction],
    "on": [MockFunction],
    "setImportInProgress": [MockFunction],
    "state": "some state",
    "updateConfiguration": [MockFunction],
    "updateState": [MockFunction],
  },
  Object {
    "clear": [MockFunction],
    "configuration": "some configuration",
    "dispose": [MockFunction],
    "emit": [MockFunction],
    "on": [MockFunction],
    "setImportInProgress": [MockFunction],
    "state": "some state",
    "updateConfiguration": [MockFunction],
    "updateState": [MockFunction],
  },
  Object {
    "clear": [MockFunction],
    "configuration": "some configuration",
    "dispose": [MockFunction],
    "emit": [MockFunction],
    "on": [MockFunction],
    "setImportInProgress": [MockFunction],
    "state": "some state",
    "updateConfiguration": [MockFunction],
    "updateState": [MockFunction],
  },
  Object {
    "clear": [MockFunction],
    "configuration": "some configuration",
    "dispose": [MockFunction],
    "emit": [MockFunction],
    "on": [MockFunction],
    "setImportInProgress": [MockFunction],
    "state": "some state",
    "updateConfiguration": [MockFunction],
    "updateState": [MockFunction],
  },
  Object {
    "clear": [MockFunction],
    "configuration": "some configuration",
    "dispose": [MockFunction],
    "emit": [MockFunction],
    "on": [MockFunction],
    "setImportInProgress": [MockFunction],
    "state": "some state",
    "updateConfiguration": [MockFunction],
    "updateState": [MockFunction],
  },
  Object {
    "clear": [MockFunction],
    "configuration": "some configuration",
    "dispose": [MockFunction],
    "emit": [MockFunction],
    "on": [MockFunction],
    "setImportInProgress": [MockFunction],
    "state": "some state",
    "updateConfiguration": [MockFunction],
    "updateState": [MockFunction],
  },
  Object {
    "clear": [MockFunction],
    "configuration": "some configuration",
    "dispose": [MockFunction],
    "emit": [MockFunction],
    "on": [MockFunction],
    "setImportInProgress": [MockFunction],
    "state": "some state",
    "updateConfiguration": [MockFunction],
    "updateState": [MockFunction],
  },
  "marquee.jsonImport",
  "marquee.jsonExport",
]
`;
